ompi/mca/coll/han/coll_han_allgather.c

```diff
@@ -0,0 +1,341 @@
+    args->up_comm = up_comm;
+    args->low_comm = low_comm;
+    args->w_rank = w_rank;
+    args->noop = noop;
+    args->is_mapbycore = is_mapbycore;
+    args->topo = topo;
+    args->req = req;
+}
+
+int
+mca_coll_han_allgather_intra(const void *sbuf, int scount,
+                             struct ompi_datatype_t *sdtype,
+                             void *rbuf, int rcount,
+                             struct ompi_datatype_t *rdtype,
+                             struct ompi_communicator_t *comm,
+                             mca_coll_base_module_t * module)
+{
+    /* Create the subcommunicators */
+    mca_coll_han_module_t *han_module = (mca_coll_han_module_t *) module;
+    mca_coll_han_comm_create_new(comm, han_module);
+    ompi_communicator_t *low_comm = han_module->sub_comm[INTRA_NODE];
+    ompi_communicator_t *up_comm = han_module->sub_comm[INTER_NODE];
+    int low_rank = ompi_comm_rank(low_comm);
+    int w_rank = ompi_comm_rank(comm);
+
+    ompi_request_t *temp_request = NULL;
+    /* Set up request */
+    temp_request = OBJ_NEW(ompi_request_t);
+    temp_request->req_state = OMPI_REQUEST_ACTIVE;
+    temp_request->req_type = OMPI_REQUEST_COLL;
+    temp_request->req_free = han_request_free;
+    temp_request->req_status = (ompi_status_public_t){0};
+    temp_request->req_complete = REQUEST_PENDING;
+
+    /* Init topo */
+    int *topo = mca_coll_han_topo_init(comm, han_module, 2);
+
+    int root_low_rank = 0;
+    /* Create lg (lower level gather) task */
+    mca_coll_task_t *lg = OBJ_NEW(mca_coll_task_t);
+    /* Setup lg task arguments */
+    mca_coll_han_allgather_t *lg_args = malloc(sizeof(mca_coll_han_allgather_t));
+    mca_coll_han_set_allgather_args(lg_args, lg, (char *) sbuf, NULL, scount, sdtype, rbuf, rcount,
+                                    rdtype, root_low_rank, up_comm, low_comm, w_rank,
+                                    low_rank != root_low_rank, han_module->is_mapbycore, topo,
+                                    temp_request);
+    /* Init and issue lg task */
+    init_task(lg, mca_coll_han_allgather_lg_task, (void *) (lg_args));
+    issue_task(lg);
+
+    ompi_request_wait(&temp_request, MPI_STATUS_IGNORE);
+
+    return OMPI_SUCCESS;
+}
+
+/* lg: lower level (shared memory) gather task */
+int mca_coll_han_allgather_lg_task(void *task_args)
+{
+    mca_coll_han_allgather_t *t = (mca_coll_han_allgather_t *) task_args;
+    char *tmp_buf = NULL, *tmp_rbuf = NULL;
+
+    OPAL_OUTPUT_VERBOSE((30, mca_coll_han_component.han_output, "[%d] HAN Allgather:  lg\n",
+                         t->w_rank));
+
+    /* If the process is one of the node leader */
+    if (!t->noop) {
+        int low_size = ompi_comm_size(t->low_comm);
+        ptrdiff_t rsize, rgap = 0;
+        rsize = opal_datatype_span(&t->rdtype->super, (int64_t) t->rcount * low_size, &rgap);
+        tmp_buf = (char *) malloc(rsize);
+        tmp_rbuf = tmp_buf - rgap;
+    }
+    /* Shared memory gather */
```

Comment:
Maybe bit picky... for a dynamic configuration this may not be a shared memory operation